prometheus/client_test.go

```diff
@@ -1,51 +1,231 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	client.inject(api)
+	return client, api, nil
+}
+
+func setupExternal() (*Client, error) {
 	conf := config.NewConfig()
 	conf.PrometheusServiceURL = "http://prometheus-istio-system.127.0.0.1.nip.io"
 	config.Set(conf)
+	return NewClient()
 }
 
-// For now it's just a runnable manual test rather than an itest.
-// To be transformed in itest when we have a test environment ready.
-// func TestGetSourceServices(t *testing.T) {
-// 	setupConfig()
-// 	prometheusClient, err := NewClient()
-// 	if err != nil {
-// 		t.Error(err)
-// 		return
-// 	}
-// 	fmt.Printf("Dependencies: \n")
-// 	incomeServices, err := prometheusClient.GetSourceServices("tutorial", "preference")
-// 	if err != nil {
-// 		t.Error(err)
-// 		return
-// 	}
-// 	for dest, origin := range incomeServices {
-// 		fmt.Printf("To: %s, From: %s \n", dest, origin)
-// 	}
-// }
-
-// For now it's just a runnable manual test rather than an itest.
-// To be transformed in itest when we have a test environment ready.
-// func TestGetServiceMetrics(t *testing.T) {
-// 	setupConfig()
-// 	prometheusClient, err := NewClient()
-// 	if err != nil {
-// 		t.Error(err)
-// 		return
-// 	}
-// 	fmt.Printf("Metrics: \n")
-// 	metrics, err := prometheusClient.GetServiceMetrics("tutorial", "preference", "5m")
-// 	if err != nil {
-// 		t.Error(err)
-// 		return
-// 	}
-// 	for desc, metric := range metrics {
-// 		fmt.Printf("Description: %s, Metric: %v \n", desc, metric)
-// 	}
-// }
+func TestGetSourceServices(t *testing.T) {
+	rqCustV1 := model.Metric{
+		"__name__":            "istio_request_count",
+		"instance":            "172.17.0.6:42422",
+		"job":                 "istio-mesh",
+		"response_code":       "200",
+		"source_service":      "customer.tutorial.svc.cluster.local",
+		"source_version":      "v1",
+		"destination_service": "preference.tutorial.svc.cluster.local",
+		"destination_version": "v1"}
+	rqCustV2 := model.Metric{
+		"__name__":            "istio_request_count",
+		"instance":            "172.17.0.6:42422",
+		"job":                 "istio-mesh",
+		"response_code":       "200",
+		"source_service":      "customer.tutorial.svc.cluster.local",
+		"source_version":      "v2",
+		"destination_service": "preference.tutorial.svc.cluster.local",
+		"destination_version": "v1"}
+	rqCustV2ToV2 := model.Metric{
+		"__name__":            "istio_request_count",
+		"instance":            "172.17.0.6:42422",
+		"job":                 "istio-mesh",
+		"response_code":       "200",
+		"source_service":      "customer.tutorial.svc.cluster.local",
+		"source_version":      "v2",
+		"destination_service": "preference.tutorial.svc.cluster.local",
+		"destination_version": "v2"}
+	vector := model.Vector{
+		&model.Sample{
+			Metric: rqCustV1,
+			Value:  4},
+		&model.Sample{
+			Metric: rqCustV2,
+			Value:  1},
+		&model.Sample{
+			Metric: rqCustV2ToV2,
+			Value:  2}}
+
+	client, api, err := setupMocked()
+	if err != nil {
+		t.Error(err)
```

Comment:
Would it be better to use an assert statement here as well? `assert.NotNil(t, err) ` ?

Reply:
Hmm... I guess you meant `assert.Nil(t, err)` instead of the "if" block?
I'll do it.

Reply:
Actually I meant `NotNil` as you want to assert that err is not `nil`.
See also https://godoc.org/github.com/stretchr/testify/assert#NotNil

Reply:
No, I don't want to assert it is not nil: I want to report it in case it is not nil, but the normal behaviour would be err == nil.

Reply:
Yes you are right. I had a brain fart

Reply:
:-)